controllers/apply-next/form-router.js

```diff
@@ -554,6 +434,125 @@ function initFormRouter(formModel) {
         }
     });
 
+    function renderStepFor(sectionSlug, stepNumber) {
+        return function(req, res, data, errors = []) {
+            const form = formBuilder({
+                locale: req.i18n.getLocale(),
+                data: data
+            });
+
+            const sectionIndex = findIndex(form.sections, s => s.slug === sectionSlug);
+            const section = form.sections[sectionIndex];
+
+            if (section) {
+                const stepIndex = parseInt(stepNumber, 10) - 1;
+                const step = section.steps[stepIndex];
+
+                if (step) {
+                    res.locals.breadcrumbs = concat(
+                        res.locals.breadcrumbs,
+                        { label: section.title, url: `${req.baseUrl}/${section.slug}` },
+                        { label: `${step.title} (Step ${stepNumber} of ${section.steps.length})` }
+                    );
+                    const { nextUrl, previousUrl } = nextAndPrevious({
+                        baseUrl: req.baseUrl,
+                        sections: form.sections,
+                        currentSectionIndex: sectionIndex,
+                        currentStepIndex: stepIndex
+                    });
+                    if (step.isRequired) {
+                        res.render(path.resolve(__dirname, './views/step'), {
+                            previousUrl,
+                            nextUrl,
+                            title: `${step.title} | ${form.title}`,
+                            csrfToken: req.csrfToken(),
+                            step: step,
+                            errors: errors
+                        });
+                    } else {
+                        res.redirect(nextUrl);
+                    }
+                } else {
+                    res.redirect(`${req.baseUrl}/${section.slug}/1`);
+                }
+            } else {
+                res.redirect(req.baseUrl);
+            }
+        };
+    }
+
+    /**
+     * Routes: Form sections
+     */
+    router
```

Comment:
Outside of the form model structure this is the other larger change. Rather than looping through the steps we instead use route parameters to look things up. We need to do this as we're always determining the full form model using `formBuilder` based on the current data (if any).

This doesn't change a huge amount of the router code but does mean we need to do some extra work to lookup the active section and step based on the current form model.